mlir-sycl/lib/Dialect/SYCL/Transforms/ConstantPropagation.cpp

```diff
@@ -831,6 +867,33 @@ void ConstantAccessorArg::propagate(OpBuilder &builder, Region &region) {
   }
 }
 
+//===----------------------------------------------------------------------===//
+// ConstantArrayArg
+//===----------------------------------------------------------------------===//
+
+void ConstantArrayArg::propagate(OpBuilder &builder, Region &region) {
+  unsigned index = getIndex();
+  Value toReplace = region.getArgument(index);
+  OpBuilder globalCloner(region.getParentOfType<gpu::GPUFuncOp>());
+  auto newGlobal =
+      globalCloner.cloneWithoutRegions(getDefiningOp<LLVM::GlobalOp>());
+  LLVM_DEBUG({
+    llvm::dbgs().indent(2) << "Handling constant array at pos #" << index
+                           << "\n";
+    llvm::dbgs().indent(4) << "Replacing using new global " << newGlobal
+                           << "\n";
+  });
+  Location loc = toReplace.getLoc();
+  Value addressof = builder.create<LLVM::AddressOfOp>(loc, newGlobal);
+  auto pt = cast<LLVM::LLVMPointerType>(toReplace.getType());
+  if (pt != addressof.getType()) {
+    LLVM_DEBUG(llvm::dbgs().indent(4) << "Performing cast to " << pt << "\n");
+    addressof = builder.create<LLVM::BitcastOp>(loc, pt, addressof);
+  }
```

Comment:
Can we guard this so we would not insert `bitcast` if targeting opaque pointers?

Reply:
I added a `cast`. We're expecting private address space pointers there, so even the address space should be the same